packages/core/lib/commands/call/run.js

```diff
@@ -0,0 +1,167 @@
+
+  const [contractNameOrAddress, functionNameOrSignature, ...args] = config._;
+  let functionEntry, transaction;
+
+  const { encoder, decoder } = config.fetchExternal
+    ? await sourceFromExternal(contractNameOrAddress, config)
+    : await sourceFromLocal(contractNameOrAddress, config);
+
+  ({ abi: functionEntry, tx: transaction } = await encoder.encodeTransaction(
+    functionNameOrSignature,
+    args
+  ));
+
+  if (functionEntry.stateMutability !== "view") {
+    console.log(
+      `WARNING!!! Not a view function\n  ABI stateMutability: ${functionEntry.stateMutability}`
+    );
+  }
+
+  const provider = new Encoder.ProviderAdapter(config.provider);
+  let result;
+  try {
+    result = await provider.request({
+      method: "eth_call",
+      params: [
+        {
+          from: config.from,
+          to: transaction.to,
+          data: transaction.data
+        },
+        config.blockNumber
+      ]
+    });
+  } catch (error) {
+    console.log(`Error Message: ${error.message}\nError Code: ${error.code}`);
+  }
+
+  if (result !== null && result !== undefined) {
+    const [decoding] = await decoder.decodeReturnValue(functionEntry, result);
+
+    for (const { value: result } of decoding.arguments) {
+      console.log(
+        util.inspect(new Codec.Export.ResultInspector(result), {
+          colors: true,
+          depth: null,
+          maxArrayLength: null,
+          breakLength: 79
+        })
+      );
+    }
+  }
+
+  async function sourceFromLocal(contractNameOrAddress, config) {
+    let encoder, decoder;
+    const contractNames = fs
+      .readdirSync(config.contracts_build_directory)
+      .filter(filename => filename.endsWith(".json"))
+      .map(filename => filename.slice(0, -".json".length));
+
+    const contracts = contractNames
+      .map(contractName => ({
+        [contractName]: config.resolver.require(contractName)
+      }))
+      .reduce((a, b) => ({ ...a, ...b }), {});
+
+    const isEmpty = Object.keys(contracts).length === 0;
+    if (isEmpty) {
+      throw new Error(
+        "No artifacts found! Please run `truffle compile` first to compile your contracts"
+      );
+    }
+
+    if (contractNameOrAddress.startsWith("0x")) {
+      const projectInfo = {
+        artifacts: Object.values(contracts)
+      };
+
+      ({ encoder, decoder } = await getEncoderDecoderForContractAddress(
+        contractNameOrAddress,
+        projectInfo
+      ));
+    } else {
+      const settings = {
+        provider: config.provider,
+        projectInfo: {
+          artifacts: Object.values(contracts)
+        }
+      };
+
+      const contract = contracts[contractNameOrAddress];
+      // Error handling to remind users to run truffle migrate first
+      const instance = await contract.deployed();
+      encoder = await Encoder.forContractInstance(instance, settings);
+      decoder = await Decoder.forContractInstance(instance, settings);
+    }
+    return { encoder, decoder };
```

Comment:
This is a really minor stylistic note, but I think this code would be easier to follow if each branch returned separately, rather than having each branch setting `{ encoder, decoder }` and having that returned at the end.  E.g., in the first case, you could just write `return await ...` and never need to explicitly set `encoder` and `decoder` at all. :)

Reply:
Yeah! I guess removing unnecessary variable declaration makes sense. Thanks @haltman-at .